doc/sphinx/proof-engine/ltac2.rst

```diff
@@ -659,168 +734,476 @@ insert in a concise way an Ltac2 variable of type :n:`constr` into a Coq term.
 Match over terms
 ~~~~~~~~~~~~~~~~
 
-Ltac2 features a construction similar to Ltac1 :n:`match` over terms, although
+Ltac2 features a construction similar to Ltac1 :tacn:`match` over terms, although
 in a less hard-wired way.
 
-.. productionlist:: coq
-   ltac2_term       : match! `ltac2_term` with `constrmatching` .. `constrmatching` end
-                    : lazy_match! `ltac2_term` with `constrmatching` .. `constrmatching` end
-                    : multi_match! `ltac2_term` with `constrmatching` .. `constrmatching` end
-   constrmatching  : | `constrpattern` => `ltac2_term`
-   constrpattern   : `term`
-                   : context  [ `term` ]
-                   : context `lident` [ `term` ]
-
-This construction is not primitive and is desugared at parsing time into
-calls to term matching functions from the `Pattern` module. Internally, it is
-implemented thanks to a specific scope accepting the :n:`@constrmatching` syntax.
-
-Variables from the :n:`@constrpattern` are statically bound in the body of the branch, to
+.. tacn:: @ltac2_match_key @ltac2_expr__term with @ltac2_match_list end
+   :name: lazy_match!; match!; multi_match!
+
+   .. insertprodn ltac2_match_key ltac2_match_pattern
+
+   .. prodn::
+      ltac2_match_key ::= lazy_match!
+      | match!
+      | multi_match!
+      ltac2_match_list ::= {? %| } {+| @ltac2_match_rule }
+      ltac2_match_rule ::= @ltac2_match_pattern => @ltac2_expr
+      ltac2_match_pattern ::= @cpattern
+      | context {? @ident } [ @cpattern ]
+
+   Evaluates :n:`@ltac2_expr__term`, which must yield a term, and matches it
+   sequentially with the :token:`ltac2_match_pattern`\s, which may have
+   metavariables.  When a match is found, metavariable values are substituted
+   into :n:`@ltac2_expr`, which is then applied.
+
+   Matching may continue depending on whether  `lazy_match!`, `match!` or `multi_match!`
+   is specified.
+
+   .. todo PR I see ? is sometimes need in the ltac2_expr and sometimes not, but
+      I don't when which is required
```

Comment:
Examples?

Reply:
Consider these examples in the text.  In the first one, 'h' is matched and its value is used.  It doesn't work with `?h`.  The second one only works with `?t` but not `t`.

```
      Goal forall A B : Prop, A -> B -> (A->B).
      intros.
      match! goal with
      | [ h : _ |- _ ] => let h := Control.hyp h in print (of_constr h); apply $h
      end.

      Ltac2 f2 t := match! t with
                    | context [ (~ ?t) ] => print (of_constr t); fail
                    | _ => ()
                    end.
```

Reply:
And the '$h' form is needed to access the value in tactic expressions but not otherwise.

Reply:
The rule is that hypothesis names are bound without needing to be preceded by a `?`.  (Even in Ltac1, locations that bind only names never need a `?`, for example `let x := _ in _` binds `x`, while `let _ := x in _` does not.  Ltac2 does not support binding names like this in terms, so the only place where this rule shows up is when matching hypothesis names.)  Unification variables in terms/types require `?` to distinguish the unification variable from potential globals of the same name.

> And the '$h' form is needed to access the value in tactic expressions but not otherwise.

That's correct.  The `$` is used for antiquotations, and this is not specific to variables bound via matching.  That is, the need to use `$` is orthogonal from whether the variable was bound via `?` or whether it was bound via an Ltac2 `let`.

Reply:
Antiquotations (the `$h` form) only exist in some notation scopes, like e.g. `constr`, `ident`, and more complex ones built out of those. Since `apply` is a notation that parses a `constr` as an argument, writing `h` there means the term variable `h` while `$h` is the Ltac2 bound variable `h` from the context.

Reply:
> The rule is that hypothesis names are bound without needing to be preceded by a `?`.

Yeah, and I recall now that Ltac1 does that, too.  OTOH, wouldn't it be more consistent to always require `?`, even for hypothesis names?

> @ppderot Antiquotations (the $h form) only exist in some notation scopes, like e.g. constr, ident, and more complex ones built out of those.

Is that the full list of notation scopes suitable for dropping into the chapter somewhere?

Reply:
> Yeah, and I recall now that Ltac1 does that, too. OTOH, wouldn't it be more consistent to always require `?`, even for hypothesis names?

Maybe... there are a couple of issues with this, though.  The first issue is that this would suggest that something like `?h : ?t, ?p : ?h = ?h |- _` would bind `t` to the type of `h`, but this is not the case, as there is check that terms bound in patterns match bound hypotheses.  The second issue is that this would require changing a great deal of code to add an extra character in many places (at least if this change were required in Ltac1 as well).  I'm not convinced, but feel free to open an issue about it.

Reply:
I'll let it be for now.  Ltac1 has other things that would be a higher priority to improve.